### PR TITLE
Change mypy for pyright

### PR DESCRIPTION
**Pull request by kalharko**
Change type checking tool, improves code quality guaranties.

### CODE REVIEW

File: mfai/pytorch/lightning_modules/clip.py [hunk 1]
```diff
@@ -72,6 +77,7 @@ def __init__(
         self.learning_rate = learning_rate
         self.min_learning_rate = min_learning_rate
         self.lr_scheduler_interval = lr_scheduler_interval
+        self.skill_score = CLIPAccuracySkillScore(top_k=1, batch_size=1)
```
Comment (tourniert) [inline mfai/pytorch/lightning_modules/clip.py:80]: Why did you add a new metric in this PR ?

File: mfai/pytorch/models/llms/llama2.py [hunk 2]
```diff
@@ -130,7 +133,10 @@ def __init__(
         )
         self.register_buffer("cos", cos)
         self.register_buffer("sin", sin)
+        self.cos = cos
+        self.sin = sin
```
Comment (tourniert) [inline mfai/pytorch/models/llms/llama2.py:137]: Why did you set this new attributes ? I don't see where they are called

File: mfai/pytorch/models/llms/llama3.py [hunk 3]
```diff
@@ -314,6 +317,7 @@ def forward_vectors(
                 diagonal=1,
             )
 
+        x = embeddings
```
Comment (tourniert) [inline mfai/pytorch/models/llms/llama3.py:320]: Why adding this line ?

File: mfai/pytorch/models/llms/qwen3_5.py [hunk 4]
```diff
@@ -22,8 +22,10 @@
 from huggingface_hub import snapshot_download
 from safetensors.torch import load_file
 from torch import Tensor
+from typing_extensions import override
 
 from mfai.pytorch.models.base import ModelType
+from mfai.tokenizers import Qwen3_5Tokenizer
```
Comment (tourniert) [inline mfai/pytorch/models/llms/qwen3_5.py:28]: Is this line will provoke to install `mfai[llm]`even when importing a basic `UNet` ?

File: mfai/pytorch/models/nlam/__init__.py [hunk 5]
```diff
@@ -67,7 +69,12 @@ class BaseGraphModel(BaseModel):
     hierarchical = False
     onnx_supported = False
     supported_num_spatial_dims = (1,)
-    num_spatial_dims: int = 1
+
+    @property
+    @override
+    def num_spatial_dims(self) -> int:
+        return 1
+
```
Comment (tourniert) [inline mfai/pytorch/models/nlam/__init__.py:77]: I do not understand why the old lines should be replaced

File: mfai/pytorch/models/deeplabv3.py [hunk 6]
```diff
@@ -298,7 +305,11 @@ class DeepLabV3(BaseModel, AutoPaddingModel):
     supported_num_spatial_dims = (2,)
     features_last: bool = False
     model_type: ModelType = ModelType.CONVOLUTIONAL
-    num_spatial_dims: int = 2
+
+    @property
+    @override
+    def num_spatial_dims(self) -> int:
+        return 2
```
Comment (tourniert) [inline mfai/pytorch/models/deeplabv3.py:312]: Same question here

File: mfai/pytorch/models/half_unet.py [hunk 7]
```diff
@@ -67,10 +70,14 @@ class HalfUNet(BaseModel, AutoPaddingModel):
...
     supported_num_spatial_dims: tuple[int, ...] = (2,)
-    num_spatial_dims: int = 2
     features_last: bool = False
     model_type: ModelType = ModelType.CONVOLUTIONAL
 
+    @property
+    @override
+    def num_spatial_dims(self) -> int:
+        return 2
+
```
Comment (tourniert) [inline mfai/pytorch/models/half_unet.py:80]: Idem

File: mfai/pytorch/models/identity.py [hunk 8]
```diff
@@ -20,13 +21,21 @@ class IdentityModel(BaseModel):
...
     onnx_supported = False
     supported_num_spatial_dims = (2, 3)
     features_last = False
     model_type = ModelType.IDENTITY
-    num_spatial_dims: int = 2
+
+    @property
+    @override
+    def num_spatial_dims(self) -> int:
+        return 2
```
Comment (tourniert) [inline mfai/pytorch/models/identity.py:38]: Idem

File: mfai/pytorch/models/resnet.py [hunk 9]
```diff
@@ -39,15 +40,17 @@ def __init__(self, out_channels: tuple[int, ...], depth: int = 5, **kwargs: Any)
             self.layer4,
         ]
 
-    def forward(self, x: Tensor) -> list[Tensor]:
+    @override
+    def forward(self, x: Tensor) -> list[Tensor]:  # type: ignore[override]
```
Comment (tourniert) [inline mfai/pytorch/models/resnet.py:44]: Why is the [ignore] being ignored when it's right above?

File: mfai/pytorch/lightning_modules/clip.py (hunk 1 above)
Comment (kalharko) [reply mfai/pytorch/lightning_modules/clip.py:80]: This is not a new metric, it is already defined in the setup method, but class members must be defined in its body or __init__ method. Otherwise, class methods or outside users have no guaranty that the member is defined.

In this case, in `validation_step`, you have no guaranty that `setup` has been called before, and `self.skill_score` could be undefined.

File: mfai/pytorch/models/llms/llama2.py (hunk 2 above)
Comment (kalharko) [reply mfai/pytorch/models/llms/llama2.py:137]: It is to help the type checker see that `self.cos` and `self.sin` are constructor-initialized attributes. Otherwise, pyright does not undersant that the `self.register_buffer` method does it.

This does not change the class behaviour.

File: mfai/pytorch/models/llms/llama3.py (hunk 3 above)
Comment (kalharko) [reply mfai/pytorch/models/llms/llama3.py:320]:
```py
x = embeddings
for i, block in enumerate(self.trf_blocks):
    x = # definition of x

x = self.final_norm(x)  # Possibly unbound variable
```

If `self.trf_blocks` is empty, we would not enter the loop and x would be undefined.

File: mfai/pytorch/models/llms/qwen3_5.py (hunk 4 above)
Comment (kalharko) [reply mfai/pytorch/models/llms/qwen3_5.py:28]:
```sh
$ uv sync #  sync without extras
$ uv run python
Python 3.11.13 (main, Aug 21 2025, 00:00:00) [GCC 11.5.0 20240719 (Red Hat 11.5.0-11)] on linux
Type "help", "copyright", "credits" or "license" for more information.
Ctrl click to launch VS Code Native REPL
>>> import mfai
>>> from mfai.pytorch.models.unet import *
```

Works.

File: mfai/pytorch/models/nlam/__init__.py (hunk 5 above)
Comment (kalharko) [reply mfai/pytorch/models/nlam/__init__.py:77]: The definition of the parent class `ModelABC` changed to have the `num_spatial_dims` be a property since it could be a computed value, instead of a class constant.

To be complete, the type checker asks that the implementation of a property be a property, not just a class member.

File: mfai/pytorch/models/deeplabv3.py (hunk 6 above)
Comment (kalharko) [reply mfai/pytorch/models/deeplabv3.py:312]: Same answer here

File: mfai/pytorch/models/half_unet.py (hunk 7 above)
Comment (kalharko) [reply mfai/pytorch/models/half_unet.py:80]: Idem

File: mfai/pytorch/models/identity.py (hunk 8 above)
Comment (kalharko) [reply mfai/pytorch/models/identity.py:38]: Idem

File: mfai/pytorch/models/resnet.py (hunk 9 above)
Comment (kalharko) [reply mfai/pytorch/models/resnet.py:44]: The signature of this method in the parent class is `forward(self, x: Tensor) -> Tensor`.
Since we change this signature to return a `list[Tensor]` in the implementation. This raises a type hinting error. Were not supposed to change the signature of a method when overriding it.

**Review by tourniert — APPROVED** [review]

File: mfai/pytorch/models/nlam/create_mesh.py [hunk 10]
```diff
@@ -135,7 +141,12 @@ def mk_2d_graph(xy: np.ndarray, nx: int, ny: int) -> networkx.DiGraph:
     return dg
 
 
-def save_edges(graph: pyg.data.Data, name: str, base_path: Path) -> None:
+def save_edges(graph: pyg_data.Data, name: str, base_path: Path) -> None:
+    assert (
+        graph.edge_index is not None
+        and graph.len is not None
+        and graph.vdiff is not None
+    )
```
Comment (LBerth) [inline mfai/pytorch/models/nlam/create_mesh.py:149]: is it okay to use assert outside of unit tests ?

File: mfai/pytorch/models/nlam/create_mesh.py [hunk 11]
```diff
@@ -1,12 +1,17 @@
...
 import matplotlib
+import matplotlib.collections
 import matplotlib.pyplot as plt
 import networkx
 import numpy as np
 import torch
-import torch_geometric as pyg
-from networkx.classes.reportviews import NodeView
+import torch_geometric.data as pyg_data
+import torch_geometric.utils as pyg_utils
```
Comment (LBerth) [inline mfai/pytorch/models/nlam/create_mesh.py:11]: Why import this way instead of just pyg ?

File: mfai/pytorch/models/nlam/create_mesh.py (hunk 10 above)
Comment (kalharko) [reply mfai/pytorch/models/nlam/create_mesh.py:149]: In this case, I do not see other way to tell the type hinter that we are certain that the graph's content is defined.

File: mfai/pytorch/models/nlam/create_mesh.py (hunk 11 above)
Comment (kalharko) [reply mfai/pytorch/models/nlam/create_mesh.py:11]: Otherwise I get the pyright warninig: `"data" is not a known attribute of module "torch_geometric" (reportAttributeAccessIssue)`

I guess `torch_geometric/__init__.py` does not include its `data` submodule in its `__all__` variable. With this syntax, the typed `data submodule` is imported and pyright can access its type hinting informations.

File: pyproject.toml [hunk 12]
```diff
@@ -107,7 +105,7 @@ Homepage = "https://github.com/meteofrance/mfai"
 # Enforces the use of prebuilt binary wheels over source distributions
 # that require a build step and executes arbitrary code when installing.
 pip.only-binary = [":all:"]
-exclude-newer = "15 days"
+# exclude-newer = "15 days"
```
Comment (LBerth) [inline pyproject.toml:108]: why remove the exclude newer ?

Comment (kalharko) [reply pyproject.toml]: I removed it, and pined pyright to an older version (no big deal)

**Review by LBerth — APPROVED** [review]
